abe/fame.go

```diff
@@ -0,0 +1,343 @@
+	pow2.Mod(pow2, a.P)
+
+	k0 := [3]*bn256.G2{new(bn256.G2).ScalarBaseMult(pow0),
+		new(bn256.G2).ScalarBaseMult(pow1),
+		new(bn256.G2).ScalarBaseMult(pow2)}
+
+	a0Inv := new(big.Int).ModInverse(sk.partInt[0], a.P)
+	a1Inv := new(big.Int).ModInverse(sk.partInt[1], a.P)
+	aInv := [2]*big.Int{a0Inv, a1Inv}
+
+	k := make([][3]*bn256.G1, len(gamma))
+	attribToI := make(map[int]int)
+	for i, y := range gamma {
+		k[i] = [3]*bn256.G1{new(bn256.G1), new(bn256.G1), new(bn256.G1)}
+		gSigma := new(bn256.G1).ScalarBaseMult(sigma[i])
+		for t := 0; t < 2; t++ {
+			hs0, err := bn256.HashG1(strconv.Itoa(y) + " 0 " + strconv.Itoa(t))
+			if err != nil {
+				return nil, err
+			}
+			hs0.ScalarMult(hs0, pow0)
+			hs1, err := bn256.HashG1(strconv.Itoa(y) + " 1 " + strconv.Itoa(t))
+			if err != nil {
+				return nil, err
+			}
+			hs1.ScalarMult(hs1, pow1)
+			hs2, err := bn256.HashG1(strconv.Itoa(y) + " 2 " + strconv.Itoa(t))
+			if err != nil {
+				return nil, err
+			}
+			hs2.ScalarMult(hs2, pow2)
+
+			k[i][t].Add(hs0, hs1)
+			k[i][t].Add(k[i][t], hs2)
+			k[i][t].Add(k[i][t], gSigma)
+			k[i][t].ScalarMult(k[i][t], aInv[t])
+		}
+
+		k[i][2].ScalarBaseMult(sigma[i])
+		k[i][2].Neg(k[i][2])
+
+		attribToI[y] = i
+	}
+
+	sigmaPrime, err := sampler.Sample()
+	if err != nil {
+		return nil, err
+	}
+	gSigmaPrime := new(bn256.G1).ScalarBaseMult(sigmaPrime)
+
+	kPrime := [3]*bn256.G1{new(bn256.G1), new(bn256.G1), new(bn256.G1)}
+	for t := 0; t < 2; t++ {
+		hs0, err := bn256.HashG1("0 0 0 " + strconv.Itoa(t))
+		if err != nil {
+			return nil, err
+		}
+		hs0.ScalarMult(hs0, pow0)
+		hs1, err := bn256.HashG1("0 0 1 " + strconv.Itoa(t))
+		if err != nil {
+			return nil, err
+		}
+		hs1.ScalarMult(hs1, pow1)
+		hs2, err := bn256.HashG1("0 0 2 " + strconv.Itoa(t))
+		if err != nil {
+			return nil, err
+		}
+		hs2.ScalarMult(hs2, pow2)
+
+		kPrime[t].Add(hs0, hs1)
+		kPrime[t].Add(kPrime[t], hs2)
+		kPrime[t].Add(kPrime[t], gSigmaPrime)
+		kPrime[t].ScalarMult(kPrime[t], aInv[t])
+		kPrime[t].Add(kPrime[t], sk.partG1[t])
+
+	}
+	kPrime[2].ScalarBaseMult(sigmaPrime)
+	kPrime[2].Neg(kPrime[2])
+	kPrime[2].Add(kPrime[2], sk.partG1[2])
+
+	return &FAMEAttribKeys{k0: k0, k: k, kPrime: kPrime, attribToI: attribToI}, nil
+}
+
+// Decrypt takes as an input a cipher and an FAMEAttribKeys and tries to decrypt
+// the cipher. This is possible only if the set of possessed attributes (and
+// corresponding keys FAMEAttribKeys) suffices the encryption policy of the
+// cipher. If this is not possible, an error is returned.
+func (a *FAME) Decrypt(cipher *FAMECipher, key *FAMEAttribKeys, pk *FAMEPubKey) (*bn256.GT, error) {
+	// find out which attributes are possessed
+	attribMap := make(map[int]bool)
+	for k := range key.attribToI {
+		attribMap[k] = true
+	}
+
+	// create a matrix of needed keys
+	preMatForKey := make([]data.Vector, 0)
+	ctForKey := make([][3]*bn256.G1, 0)
+	rowToAttrib := make([]int, 0)
+	for i := 0; i < len(cipher.msp.Mat); i++ {
```

Comment:
How many rows is the `cipher.msp.Mat` expected to have? If the number is big, maybe it would be better to pre-calculate the number of times where the `if` clause will be executed and preallocate the slices with the appropriate length. Or maybe just add the third parameter (`cap` - capacity) to the `make` calls and set it to the number of rows.

Reply:
Good question. The matrix will have as many rows as there are attributes used in the boolean formula describing the decryption policy. In a normal usage this is probably not too big, but maybe in some unexpected use could also be a bit bigger. How big the number must be so that the optimization makes sense?

Reply:
I would say at least a 100 or maybe 1000, but that's just a guess... If we don't want to waste memory (by using the capacity parameter, we'll probably allocate too much), it's probably the best if we add a loop before which calculates the size. Then we can allocate the right amount of memory and not lose too much on speed (a simple loop should be fast enough).

Reply:
I changed as suggested. Thanks.